src/passes/externalArgumentModifier.ts

```diff
@@ -0,0 +1,186 @@
+      node.vParameters.vParameters.forEach((varDecl) => {
+        if (
+          varDecl.storageLocation === DataLocation.Memory &&
+          varDecl.vType instanceof UserDefinedTypeName &&
+          varDecl.vType.vReferencedDeclaration instanceof StructDefinition
+        ) {
+          memoryStructs.push(varDecl);
+          varDecl.storageLocation = DataLocation.CallData;
+        }
+      });
+      if (memoryStructs.length > 0) {
+        const variableDeclarationStatements = memoryStructs.map((varDecl) => {
+          return this.createVariableDeclarationStatement(varDecl, ast);
+        });
+        variableDeclarationStatements.forEach((statement) => {
+          node.vBody?.insertAtBeginning(statement);
+        });
+        ast.setContextRecursive(node);
+      }
+    }
+    this.commonVisit(node, ast);
+  }
+
+  private createVariableDeclarationStatement(
+    calldataStruct: VariableDeclaration,
+    ast: AST,
+  ): VariableDeclarationStatement {
+    const memoryStruct = cloneASTNode(calldataStruct, ast);
+    memoryStruct.name = calldataStruct.name + '_mem';
+    memoryStruct.storageLocation = DataLocation.Memory;
+    this.oldToNewStructMapping.set(calldataStruct, memoryStruct);
+    const callDataVarDecl = createIdentifier(memoryStruct, ast);
+    const structDeclarationStatement = new VariableDeclarationStatement(
+      ast.reserveId(),
+      '',
+      [1],
+      [memoryStruct],
+      this.createStructConstructor(calldataStruct, memoryStruct, ast),
+    );
+    this.identifiersNotToBeChanged.push(callDataVarDecl);
+    return structDeclarationStatement;
+  }
+
+  private createStructConstructor(
+    calldataStruct: VariableDeclaration,
+    memoryStruct: VariableDeclaration,
+    ast: AST,
+  ): FunctionCall {
+    const structConstructorTypeString = this.generateStructTypeString(memoryStruct);
+    assert(
+      calldataStruct.vType instanceof UserDefinedTypeName &&
+        calldataStruct.vType.vReferencedDeclaration instanceof StructDefinition,
+    );
+
+    const structConstructor = new FunctionCall(
+      ast.reserveId(),
+      '',
+      structConstructorTypeString,
+      FunctionCallKind.StructConstructorCall,
+      this.generateStructIdentifier(calldataStruct, ast),
+      this.generateMemberAccess(calldataStruct, ast),
+    );
+    return structConstructor;
+  }
+
+  private generateStructTypeString(varDecl: VariableDeclaration): string {
+    const dataLocation = varDecl.storageLocation;
+    assert(varDecl.vType instanceof UserDefinedTypeName);
+    const structName = varDecl.typeString;
+
+    return `${structName} ${dataLocation}`;
+  }
+
+  generateStructIdentifier(varDecl: VariableDeclaration, ast: AST): Expression {
+    assert(
+      varDecl.vType instanceof UserDefinedTypeName &&
+        varDecl.vType.vReferencedDeclaration instanceof StructDefinition,
+    );
+    const funcIdentifier = new Identifier(
+      ast.reserveId(),
+      '',
+      `type(${varDecl.typeString} storage pointer)`,
```

Comment:
Do we have a function that can do this?

Reply:
CreateIdentifier is the usual goto for this, but it doesn't take into account the typestring differences for reference types or the need to wrap it in `type()`, so there's nothing currently that would produce this exact result